### PR TITLE
MH-13545 Collect and visualize published hours of video

### PR DESCRIPTION
**Pull request by pmiddend**
This PR extends the "hourly resolution" PR (#876) and...

- adds a way to write statistics from workflows.
- adds a new way to aggregate statistics, using a new running total provider.

The statistics can be written using a new workflow operation called `statistics-writer`, which basically does what it says. It can only write the duration of a track currently, but you can think of many extensions 
to this later on.

One use case for this is to collect and visualize the currently published hours of video. These statistics are then displayed in the organization statistics.

This work is sponsored by SWITCH.

### CODE REVIEW

File: docs/guides/admin/docs/configuration/admin-ui/statistics.md [hunk 1]
```diff
@@ -106,6 +107,65 @@ Here is an example json configuration for a provider which generates charts for
...
+```json
+{
+  "id": "organization.publishedhours.influx",
+  "title": "STATISTICS.TITLE.PUBLISHEDHOURS",
+  "description": "STATISTICS.DESCRIPTION.PUBLISHEDHOURS",
+  "resourceType": "ORGANIZATION",
+  "sources": [{
+    "measurement": "infinite.publishedhours",
+    "aggregation": "SUM",
+    "aggregationVariable": "hours",
```
Comment (KatrinIhler) [inline docs/guides/admin/docs/configuration/admin-ui/statistics.md:135]: The description doesn't match the JSON example here.

File: docs/guides/admin/docs/configuration/admin-ui/statistics.md [hunk 2]
```diff
@@ -106,6 +107,65 @@ Here is an example json configuration for a provider which generates charts for
...
+    <configuration key="retract">false</configuration>
+    <configuration key="measurement-name">publishedseconds</configuration>
+    <configuration key="organization-resource-id-name">organizationId</configuration>
+    <configuration key="length-field-name">hours</configuration>
+    <configuration key="temporal-resolution">hours</configuration>
+  </configurations>
+</operation>
+```
+
+To _decrement_ the video length in case of retractions, you can copy the snippet above and change the `retract`
```
Comment (KatrinIhler) [inline docs/guides/admin/docs/configuration/admin-ui/statistics.md]: I'd scratch "copy the snippet above and"

File: modules/statistics-provider-influx/src/main/java/org/opencastproject/statistics/provider/influx/StatisticsProviderInfluxService.java [hunk 3]
```diff
@@ -86,22 +93,37 @@ public void deactivate(ComponentContext cc) {
...
+                ResourceType.ORGANIZATION,
+                providerCfg.getTitle(),
+                providerCfg.getDescription(),
+                providerCfg.getSources());
+        break;
+      default:
+        break;
+    }
+    if (provider == null) {
       throw new ConfigurationException("Unknown influx statistics type: " + providerCfg.getType());
```
Comment (KatrinIhler) [inline modules/statistics-provider-influx/src/main/java/org/opencastproject/statistics/provider/influx/StatisticsProviderInfluxService.java]: Why not put this into the default case?

File: modules/statistics-provider-influx/src/main/java/org/opencastproject/statistics/provider/influx/StatisticsProviderInfluxService.java [hunk 4]
```diff
@@ -170,4 +193,49 @@ private void disconnectInflux() {
     }
   }
 
+  @Override
+  public void writeDuration(
```
Comment (KatrinIhler) [inline modules/statistics-provider-influx/src/main/java/org/opencastproject/statistics/provider/influx/StatisticsProviderInfluxService.java:192]: Wouldn't the StatisticsWriter warrant its own class class like StatisticsWriterImpl, especially if this gets extended in the future?

File: modules/statistics-writer-workflowoperation/src/main/resources/OSGI-INF/operations/statisticswriter.xml [hunk 5]
```diff
@@ -0,0 +1,19 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<scr:component xmlns:scr="http://www.osgi.org/xmlns/scr/v1.1.0"
```
Comment (KatrinIhler) [inline modules/statistics-writer-workflowoperation/src/main/resources/OSGI-INF/operations/statisticswriter.xml]: I feel like statistics-writer-woh.xml would be a better name.

File: modules/statistics-service-api/src/main/java/org/opencastproject/statistics/api/StatisticsProviderRegistry.java [hunk 6]
```diff
@@ -39,4 +39,17 @@
    * @param provider The provider to remove.
    */
   void removeProvider(StatisticsProvider provider);
+
+  /**
+   * Add a new writer instance
```
Comment (KatrinIhler) [inline modules/statistics-service-api/src/main/java/org/opencastproject/statistics/api/StatisticsProviderRegistry.java:44]: I don't find it intuitive that the ServiceProviderRegistry also handles the writers. Maybe move all of this into the general StatisticsService interface instead, or is there a need for a specific registry interface?

File: modules/statistics-writer-workflowoperation/src/main/java/org/opencastproject/workflow/handler/statisticswriter/StatisticsWriterWorkflowOperationHandler.java [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+  }
+
+  private SecurityService securityService;
+
+  public void setSecurityService(SecurityService securityService) {
+    this.securityService = securityService;
+  }
+
+  protected void activate(org.osgi.service.component.ComponentContext cc) {
+
```
Comment (KatrinIhler) [inline modules/statistics-writer-workflowoperation/src/main/java/org/opencastproject/workflow/handler/statisticswriter/StatisticsWriterWorkflowOperationHandler.java]: Do we need this?

**Review by KatrinIhler — COMMENTED** [review]
Some thoughts below.

File: modules/statistics-provider-influx/src/main/java/org/opencastproject/statistics/provider/influx/StatisticsProviderInfluxService.java (hunk 4 above)
Comment (pmiddend) [reply modules/statistics-provider-influx/src/main/java/org/opencastproject/statistics/provider/influx/StatisticsProviderInfluxService.java:192]: I get your point, but I'd like to file this under "premature abstraction". As soon as this gets extended, we'll see how this class is supposed to look like.

File: modules/statistics-service-api/src/main/java/org/opencastproject/statistics/api/StatisticsProviderRegistry.java (hunk 6 above)
Comment (pmiddend) [reply modules/statistics-service-api/src/main/java/org/opencastproject/statistics/api/StatisticsProviderRegistry.java:44]: I admit, the relationships between the statistics interfaces are a little convoluted. The thing about `StatisticsService` (and it’s raison d’etre) is, it has a remote implementation to retrieve statistics. This makes no sense for the `StatisticsProviderRegistry`, because how would you add a provider, which is a piece of Java code, remotely? So, if we were to merge `StatisticsService` and the provider registry, we’d have to throw an exception for the methods we’re not implementing, which isn’t intuitive as well.

Granted, we could add *another* interface `StatisticsWriterRegistry` and let `StatisticsServiceImpl` implement that, but I don’t see that much need for it, given that the current registry isn’t overflowing with methods to implement.

File: docs/guides/admin/docs/configuration/admin-ui/statistics.md (hunk 1 above)
Comment (pmiddend) [reply docs/guides/admin/docs/configuration/admin-ui/statistics.md:135]: Good catch. I'll change everything to "hours", since that was the original purpose.

File: docs/guides/admin/docs/configuration/admin-ui/statistics.md (hunk 2 above)
Comment (pmiddend) [reply docs/guides/admin/docs/configuration/admin-ui/statistics.md]: Fixed

File: modules/statistics-provider-influx/src/main/java/org/opencastproject/statistics/provider/influx/StatisticsProviderInfluxService.java (hunk 3 above)
Comment (pmiddend) [reply modules/statistics-provider-influx/src/main/java/org/opencastproject/statistics/provider/influx/StatisticsProviderInfluxService.java]: No reason, I'll change it.

File: modules/statistics-writer-workflowoperation/src/main/resources/OSGI-INF/operations/statisticswriter.xml (hunk 5 above)
Comment (pmiddend) [reply modules/statistics-writer-workflowoperation/src/main/resources/OSGI-INF/operations/statisticswriter.xml]: It's more consistent, yes. I'll change it.

File: modules/statistics-writer-workflowoperation/src/main/java/org/opencastproject/workflow/handler/statisticswriter/StatisticsWriterWorkflowOperationHandler.java (hunk 7 above)
Comment (pmiddend) [reply modules/statistics-writer-workflowoperation/src/main/java/org/opencastproject/workflow/handler/statisticswriter/StatisticsWriterWorkflowOperationHandler.java]: Nope, copy&paste BS. ;)

File: modules/statistics-service-api/src/main/java/org/opencastproject/statistics/api/StatisticsProviderRegistry.java (hunk 6 above)
Comment (KatrinIhler) [reply modules/statistics-service-api/src/main/java/org/opencastproject/statistics/api/StatisticsProviderRegistry.java:44]: Ah, I see now. In that case, maybe we can come up with a more general name than StatisticsProviderRegistry? Although I have to admit that I can't think of one right now...

**Comment by staubesv** [comment]
@pmiddend Please note that this PR currently has merge conflicts.

**Comment by staubesv** [comment]
@pmiddend Please note that a recent merge of PR #935 (or #934) has caused merge conflicts in this PR.

File: docs/guides/admin/docs/configuration/admin-ui/statistics.md [hunk 8]
```diff
@@ -115,6 +116,64 @@ behavior in the way that exported series statistics contain the data of all even
...
+    <configuration key="retract">false</configuration>
+    <configuration key="measurement-name">publishedhours</configuration>
+    <configuration key="organization-resource-id-name">organizationId</configuration>
+    <configuration key="length-field-name">hours</configuration>
+    <configuration key="temporal-resolution">hours</configuration>
+  </configurations>
+</operation>
+```
+
+To _decrement_ the video length in case of retractions, simply change the `retract` property to `true`.
```
Comment (gregorydlogan) [inline docs/guides/admin/docs/configuration/admin-ui/statistics.md]: I would reword this: "To _decrement_ the running total of hours in the case of retractions, set the `retract` property to `true`.  In the default case, or when the `retract` property is `false` the running total is not decremented when a retraction occurs."

File: docs/guides/admin/docs/configuration/admin-ui/statistics.md [hunk 9]
```diff
@@ -115,6 +116,64 @@ behavior in the way that exported series statistics contain the data of all even
...
+  }],
+  "type": "runningtotal"
+}
+```
+
+Note that the published hours entries can be negative, in case we retract a video.
+
+When the `runningtotal` provider is asked to report on, for example, the monthly hours of video for a specific year,
+it will first take the sum of all video lengths _up until_ that year. Then, for each month, it will take the sum of
+all the entries in that month, and add it to the previous value. And so on for the next months.
```
Comment (gregorydlogan) [inline docs/guides/admin/docs/configuration/admin-ui/statistics.md:152]: This seems like it would be a massive perf hole as the number of videos goes up, but I'm assuming SWITCH is already all over that sooo :)

File: docs/guides/admin/docs/workflowoperationhandlers/statistics-writer.md [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+InfluxDB. Currently, it only writes the length of the video in seconds to the data base. It can be configured to
+write the negative length and can thus be used for retract workflows, too.
+
+
+Parameter Table
+---------------
+
+|configuration keys            |required|description                                                        |
+|------------------------------|--------|-------------------------------------------------------------------|
+|flavor                        |yes     |The flavor of the track you want to publish statistics to          |
```
Comment (gregorydlogan) [inline docs/guides/admin/docs/workflowoperationhandlers/statistics-writer.md]: The flavor of the track you want to publish statistics about, I suspect.

File: docs/guides/admin/docs/workflowoperationhandlers/statistics-writer.md [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+|measurement-name              |yes     |Measurement name of the statistics DB                              |
```
Comment (gregorydlogan) [inline docs/guides/admin/docs/workflowoperationhandlers/statistics-writer.md]: ...name __in__ the statistics DB?

File: modules/statistics-provider-influx/src/main/java/org/opencastproject/statistics/provider/influx/provider/InfluxRunningTotalStatisticsProvider.java [hunk 12]
```diff
@@ -0,0 +1,189 @@
...
+    if (!(o instanceof Double)) {
+      throw new RuntimeException("invalid results returned for aggregation");
+    }
+    return previousSeriesTotal + (Double) o;
+  }
+
+  private double getPreviousTotal(
+          final InfluxProviderConfiguration.InfluxProviderSource source, final String resourceId, final Instant from) {
+    final Query beforeQuery = BoundParameterQuery.QueryBuilder
+            .newQuery("SELECT SUM(" + source.getAggregationVariable() + ") FROM " + source.getMeasurement() + " WHERE "
```
Comment (gregorydlogan) [inline modules/statistics-provider-influx/src/main/java/org/opencastproject/statistics/provider/influx/provider/InfluxRunningTotalStatisticsProvider.java:92]: Why did you parameterize some of these variables, and not others?

File: modules/statistics-provider-influx/src/main/java/org/opencastproject/statistics/provider/influx/provider/InfluxRunningTotalStatisticsProvider.java [hunk 13]
```diff
@@ -0,0 +1,189 @@
...
+    final String influxGrouping = dataResolutionToInfluxGrouping(resolution);
+    final List<Tuple<Instant, Instant>> periods = getPeriods(from, to, resolution, zoneId);
+    final List<String> labels = new ArrayList<>();
+    final List<Double> values = new ArrayList<>();
+    final InfluxProviderConfiguration.InfluxProviderSource source = getSource(resolution);
+    double previousTotal = getPreviousTotal(source, resourceId, from);
+    for (final Tuple<Instant, Instant> period : periods) {
+      final Query query = BoundParameterQuery.QueryBuilder
+              .newQuery("SELECT " + source.getAggregation() + "(" + source.getAggregationVariable() + ") FROM " + source
+                      .getMeasurement() + " WHERE " + source.getResourceIdName()
```
Comment (gregorydlogan) [inline modules/statistics-provider-influx/src/main/java/org/opencastproject/statistics/provider/influx/provider/InfluxRunningTotalStatisticsProvider.java:118]: Parameterization again

File: modules/statistics-service-api/src/main/java/org/opencastproject/statistics/api/StatisticsProviderRegistry.java (hunk 6 above)
Comment (gregorydlogan) [reply modules/statistics-service-api/src/main/java/org/opencastproject/statistics/api/StatisticsProviderRegistry.java:44]: StatisticsWriterRegistry is the only thing I can think of, and that's not any better :D

**Review by gregorydlogan — COMMENTED** [review]
Looks good aside from a couple of questions.

File: modules/statistics-provider-influx/src/main/java/org/opencastproject/statistics/provider/influx/provider/InfluxRunningTotalStatisticsProvider.java (hunk 12 above)
Comment (pmiddend) [reply modules/statistics-provider-influx/src/main/java/org/opencastproject/statistics/provider/influx/provider/InfluxRunningTotalStatisticsProvider.java:92]: More than one reason:

1. The parametrized values are user generated ones, preventing injections.
2. It's copy&pasted from the `TimeSeriesInfluxDbProvider` (so we'd have to change it there, risking merge conflicts.
3. I don't even know if Influx allows arbitrary placeholders.

File: modules/statistics-provider-influx/src/main/java/org/opencastproject/statistics/provider/influx/provider/InfluxRunningTotalStatisticsProvider.java (hunk 13 above)
Comment (pmiddend) [reply modules/statistics-provider-influx/src/main/java/org/opencastproject/statistics/provider/influx/provider/InfluxRunningTotalStatisticsProvider.java:118]: See above.

File: modules/statistics-service-api/src/main/java/org/opencastproject/statistics/api/StatisticsProviderRegistry.java (hunk 6 above)
Comment (pmiddend) [reply modules/statistics-service-api/src/main/java/org/opencastproject/statistics/api/StatisticsProviderRegistry.java:44]: @gregorydlogan Well, it's not just for writers.

Anybody against `StatisticsCoordinator`? It doesn't feel wise to hold up this PR because of a naming issue. At least it's not a `StatisticsProviderFactorySingletonInstanceObserver` :D

File: docs/guides/admin/docs/workflowoperationhandlers/statistics-writer.md (hunk 11 above)
Comment (pmiddend) [reply docs/guides/admin/docs/workflowoperationhandlers/statistics-writer.md]: Thanks fixed.

File: docs/guides/admin/docs/workflowoperationhandlers/statistics-writer.md (hunk 10 above)
Comment (pmiddend) [reply docs/guides/admin/docs/workflowoperationhandlers/statistics-writer.md]: Yep, sounds better, fixed.

File: docs/guides/admin/docs/configuration/admin-ui/statistics.md (hunk 8 above)
Comment (pmiddend) [reply docs/guides/admin/docs/configuration/admin-ui/statistics.md]: Sure thing, fixed.

File: modules/statistics-service-api/src/main/java/org/opencastproject/statistics/api/StatisticsProviderRegistry.java (hunk 6 above)
Comment (KatrinIhler) [reply modules/statistics-service-api/src/main/java/org/opencastproject/statistics/api/StatisticsProviderRegistry.java:44]: Fine with me.

Comment (gregorydlogan) [reply modules/statistics-service-api/src/main/java/org/opencastproject/statistics/api/StatisticsProviderRegistry.java:44]: Sounds good to me.

File: modules/statistics-provider-influx/src/main/java/org/opencastproject/statistics/provider/influx/provider/InfluxRunningTotalStatisticsProvider.java (hunk 12 above)
Comment (gregorydlogan) [reply modules/statistics-provider-influx/src/main/java/org/opencastproject/statistics/provider/influx/provider/InfluxRunningTotalStatisticsProvider.java:92]: Fair enough!

**Comment by pmiddend** [comment]
Thanks guys for the review. I've fixed the issues mentioned.
